programs/ido-program/programs/ido-program/src/lib.rs

```diff
@@ -0,0 +1,438 @@
+
+    pub rent: Sysvar<'info, Rent>,
+
+    pub clock: Sysvar<'info, Clock>,
+
+    pub system_program: Program<'info, System>,
+}
+
+impl<'info> InitializePool<'info> {
+    fn accounts(ctx: &Context<InitializePool<'info>>, bump: u8) -> Result<()> {
+        let expected_signer = Pubkey::create_program_address(
+            &[ctx.accounts.pool_native.mint.as_ref(), &[bump]],
+            ctx.program_id,
+        )
+        .map_err(|_| ErrorCode::InvalidBump)?;
+        if ctx.accounts.pool_signer.key != &expected_signer {
+            return Err(ErrorCode::InvalidBump.into());
+        }
+        Ok(())
+    }
+}
+
+#[derive(Accounts)]
+pub struct ExchangeFiatForRedeemable<'info> {
+    #[account(mut, has_one = redeemable_mint, has_one = pool_fiat)]
+    pub pool: Box<Account<'info, PoolAccount>>,
+
+    ///CHECK: This is not dangerous
+    #[account(seeds = [pool.native_mint.as_ref()], bump = pool.bump)]
+    pool_signer: AccountInfo<'info>,
+
+    #[account(
+        mut,
+        constraint = redeemable_mint.mint_authority == COption::Some(*pool_signer.key)
+    )]
+    pub redeemable_mint: Account<'info, Mint>,
+
+    #[account(mut, constraint = pool_fiat.mint == *fiat_mint.to_account_info().key)]
+    pub fiat_mint: Account<'info, Mint>,
+
+    #[account(mut, constraint = pool_fiat.owner == *pool_signer.key)]
+    pub pool_fiat: Account<'info, TokenAccount>,
+
+    #[account(mut)]
+    pub authority: Signer<'info>,
+
+    #[account(mut, constraint = investor_fiat.owner == *authority.key)]
+    pub investor_fiat: Account<'info, TokenAccount>,
+
+    #[account(mut, constraint = investor_redeemable.owner == *authority.key)]
+    pub investor_redeemable: Account<'info, TokenAccount>,
+
+    #[account(constraint = token_program.key == &token::ID)]
+    pub token_program: Program<'info, Token>,
+
+    pub clock: Sysvar<'info, Clock>,
+}
+
+#[derive(Accounts)]
+pub struct ExchangeRedeemableForNative<'info> {
+    #[account(has_one = redeemable_mint, has_one = pool_native)]
+    pub pool: Box<Account<'info, PoolAccount>>,
+
+    /// CHECK: This is not dangerous
+    #[account(seeds = [pool.native_mint.as_ref()], bump = pool.bump)]
+    pool_signer: AccountInfo<'info>,
+
+    #[account(
+        mut,
+        constraint = redeemable_mint.mint_authority == COption::Some(*pool_signer.key)
+    )]
+    pub redeemable_mint: Account<'info, Mint>,
+
+    #[account(mut, constraint = pool_native.owner == *pool_signer.key)]
+    pub pool_native: Account<'info, TokenAccount>,
+
+    #[account(mut)]
+    pub authority: Signer<'info>,
+
+    #[account(mut, constraint = investor_native.owner == *authority.key)]
+    pub investor_native: Account<'info, TokenAccount>,
+
+    #[account(mut, constraint = investor_redeemable.owner == *authority.key)]
+    pub investor_redeemable: Account<'info, TokenAccount>,
+
+    #[account(constraint = token_program.key == &token::ID)]
+    pub token_program: Program<'info, Token>,
+
+    pub clock: Sysvar<'info, Clock>,
```

Comment:
Get the `Clock` from runtime with `Clock::get()?` instead of passing as an account
```suggestion
```